File: examples/typescript/part-3/07-parameter-validation-schema.ts
// Illustrative examples for notes/part-3-processing-security/07-parameter-validation-schema.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).
/// <reference path="../types.d.ts" />

import { isAbsolute, normalize, resolve, sep } from "node:path";

const ALLOWED_HOSTS = new Set(["api.example.com", "docs.example.com"]);

// --- JSON schema через строгий decoder ---

interface FetchURLArgs {
  url: string;
  timeoutSeconds: number;
}

function decodeStrict(raw: string): Record<string, unknown> {
  const obj = JSON.parse(raw);
  if (typeof obj !== "object" || obj === null || Array.isArray(obj)) {
    throw new Error("expected a JSON object");
  }
  return obj as Record<string, unknown>;
}

function validateFetchURLArgs(obj: Record<string, unknown>): FetchURLArgs {
  const allowed = new Set(["url", "timeout_seconds"]);
  const extra = Object.keys(obj).filter((k) => !allowed.has(k));
  if (extra.length > 0) {
    throw new Error(`unknown fields: ${extra.join(", ")}`);
  }

  const url = typeof obj.url === "string" ? obj.url : "";
  if (!url) {
    throw new Error("url is required");
  }

  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw new Error("invalid url");
  }
  if (parsed.protocol !== "https:") {
    throw new Error("only https scheme is allowed");
  }
  if (!ALLOWED_HOSTS.has(parsed.hostname.toLowerCase())) {
    throw new Error(`host is not allowed: ${parsed.hostname}`);
  }

  const timeout = obj.timeout_seconds;
  if (
    typeof timeout !== "number" ||
    !Number.isInteger(timeout) ||
    timeout <= 0 ||
    timeout > 10
  ) {
    throw new Error("timeout must be between 1 and 10 seconds");
  }

  return { url, timeoutSeconds: timeout };
}

// --- Безопасная проверка пути ---

function safeJoin(baseDir: string, userPath: string): string {
  if (!userPath) {
    throw new Error("path is required");
  }

  const clean = normalize(userPath);
  if (isAbsolute(clean)) {
    throw new Error("absolute paths are not allowed");
  }

  const baseAbs = resolve(baseDir);
  const fullAbs = resolve(baseAbs, clean);

  if (fullAbs !== baseAbs && !fullAbs.startsWith(baseAbs + sep)) {
    throw new Error("path escapes base directory");
  }

  return fullAbs;
}

// --- Ownership check ---

interface ResourceAccessChecker {
  canAccessCustomer(userId: string, customerId: string): boolean;
}

function validateReadCustomer(
  userId: string,
  customerId: string,
  access: ResourceAccessChecker,
): void {
  if (!customerId) {
    throw new Error("customer_id is required");
  }
  if (!access.canAccessCustomer(userId, customerId)) {
    throw new Error("customer does not belong to user");
  }
}

export {
  decodeStrict,
  validateFetchURLArgs,
  safeJoin,
  validateReadCustomer,
};

export type { FetchURLArgs, ResourceAccessChecker };

File: examples/python/part-3/07-parameter-validation-schema.py
# Illustrative examples for notes/part-3-processing-security/07-parameter-validation-schema.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

import json
import os
from typing import Protocol
from urllib.parse import urlparse

ALLOWED_HOSTS = {"api.example.com", "docs.example.com"}


# --- JSON schema через строгий decoder ---


def decode_strict(raw: str) -> dict:
    obj = json.loads(raw)
    if not isinstance(obj, dict):
        raise ValueError("expected a JSON object")
    return obj


def validate_fetch_url_args(obj: dict) -> tuple[str, int]:
    allowed_fields = {"url", "timeout_seconds"}
    extra = set(obj) - allowed_fields
    if extra:
        raise ValueError(f"unknown fields: {sorted(extra)}")

    url = obj.get("url", "")
    if not url:
        raise ValueError("url is required")

    parsed = urlparse(url)
    if parsed.scheme != "https":
        raise ValueError("only https scheme is allowed")

    host = (parsed.hostname or "").lower()
    if host not in ALLOWED_HOSTS:
        raise ValueError(f"host is not allowed: {host}")

    timeout = obj.get("timeout_seconds", 0)
    if not isinstance(timeout, int) or timeout <= 0 or timeout > 10:
        raise ValueError("timeout must be between 1 and 10 seconds")

    return url, timeout


# --- Безопасная проверка пути ---


def safe_join(base_dir: str, user_path: str) -> str:
    if not user_path:
        raise ValueError("path is required")

    clean = os.path.normpath(user_path)
    if os.path.isabs(clean):
        raise ValueError("absolute paths are not allowed")

    base_abs = os.path.abspath(base_dir)
    full_abs = os.path.abspath(os.path.join(base_abs, clean))

    if full_abs != base_abs and not full_abs.startswith(base_abs + os.sep):
        raise ValueError("path escapes base directory")

    return full_abs


# --- Ownership check ---


class ResourceAccessChecker(Protocol):
    def can_access_customer(self, user_id: str, customer_id: str) -> bool: ...


def validate_read_customer(
    user_id: str,
    customer_id: str,
    access: ResourceAccessChecker,
) -> None:
    if not customer_id:
        raise ValueError("customer_id is required")
    if not access.can_access_customer(user_id, customer_id):
        raise ValueError("customer does not belong to user")
